contracts/resolver/ZNSStringResolver.sol

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.18;
+
+import { ERC165 } from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { IZNSStringResolver } from "./IZNSStringResolver.sol";
+import { AAccessControlled } from "../access/AAccessControlled.sol";
+import { ARegistryWired } from "../registry/ARegistryWired.sol";
+
+
+/**
+ * @title The specific Resolver for ZNS that maps domain hashes to strings.
+ */
+contract ZNSStringResolver is
+    UUPSUpgradeable,
+    AAccessControlled,
+    ARegistryWired,
+    ERC165,
+    IZNSStringResolver {
+
+    mapping(bytes32 domainHash => string resolvedString) internal domainStrings;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Initializer for the `ZNSStringResolver` proxy.
+     * Note that setter functions are used instead of direct state variable assignments
+     * to use access control at deploy time. Only ADMIN can call this function.
+     * @param accessController_ The address of the `ZNSAccessController` contract
+     * @param registry_ The address of the `ZNSRegistry` contract
+     */
+    function initialize(address accessController_, address registry_) external override initializer {
+        _setAccessController(accessController_);
+        setRegistry(registry_);
+    }
+
+    /**
+     * @dev Returns string associated with a given domain name hash.
+     * @param domainHash The identifying hash of a domain's name
+     */
+    function resolveDomainString(
+        bytes32 domainHash
+    ) external view override returns (string memory) {
+        return domainStrings[domainHash];
+    }
+
+    /**
+     * @dev Sets the string for a domain name hash.
+     * @param domainHash The identifying hash of a domain's name
+     * @param newString The new string to map the domain to
+     */
+    function setString(
+        bytes32 domainHash,
+        string calldata newString
+    ) external override {
+        // only owner or operator of the current domain can set the string
+        require(
```

Comment:
Using custom errors is more gas efficient. I know they aren't used in the other contracts but they have already been deployed, so maybe we can update that at some point but for now use them instead of `require` and do the opposite logic of what you're checking here

```solidity
if (!registry.IsOwnerOrOperator(domainHash, msg.sender)) {
    revert CustomErrorName();
}
```